posts/Gunnar/2019-01-23-testing-cdi-beans-and-persistence-layer-under-java-se.adoc

```diff
@@ -0,0 +1,456 @@
+= Testing CDI Beans and the Persistence Layer Under Java SE
+Gunnar Morling
+:awestruct-tags: [ "Discussions" ]
+:awestruct-layout: blog-post
+---
+
+When it comes to testing Java EE applications, there's a wide spectrum of tools and approaches at our disposal.
+Depending on the specific goals and requirements of a given test, options range from
+plain unit tests of single classes to comprehensive integration tests deployed into a container (e.g. via http://arquillian.org/[Arquillian])
+and driven through tools such as http://rest-assured.io/[REST Assured].
+
+In this post I'd like to discuss one testing approach which represents some kind of a middle ground:
+launching a local CDI container and a JPA runtime, connected to an in-memory database.
+That way you can test CDI beans (e.g. containing business logic) in conjunction with the persistence layer (e.g. JPA-based repositories) under plain Java SE.
+
+This allows to test individual classes and components as they interact with others
+(e.g. without mocking of repositories when testing the business logic),
+while still benefitting from fast execution times (no container management/deployment and remote API calls needed).
+The approach also allows tests around services our application might rely on, such as interceptors, events, transactional semantics and other things
+that'd otherwise require deployment into a container.
+Finally, these tests are easy to debug, as everything runs in the local VM and no remote processes are involved.
+
+To make the approach worthwhile, the following requirements should be satisfied by the testing infrastructure:
+
+* Allow to obtain CDI beans via dependency injection, with support for all the CDI goodness such as interceptors, decorators, events etc.
```

Comment:
Recently I was told that [`allow`](https://www.merriam-webster.com/dictionary/allow) with the meaning of "permit" is a [transitive verb](https://www.merriam-webster.com/dictionary/transitive), i.e. requires a direct object: "Allows **you/developers/etc** to obtain...". Needless to say that I used it wrong for years ;-)

Reply:
Interesting, thanks. Changing it.